docs/reference/search/knn-search.asciidoc

```diff
@@ -0,0 +1,129 @@
+[[knn-search]]
+=== kNN search API
+++++
+<titleabbrev>kNN search</titleabbrev>
+++++
+
+experimental::[]
+
+Performs a k-nearest neighbor search and returns the matching documents.
+
+[source,console]
+----
+PUT my-vector-index
+{
+  "mappings": {
+    "properties": {
+      "image_vector": {
+        "type": "dense_vector",
+        "dims": 3,
+        "index": true,
+        "similarity": "l2_norm"
+      }
+    }
+  }
+}
+
+GET /my-vector-index/_knn_search
+{
+  "knn": {
+    "field": "image_vector",
+    "query_vector": [0.3, 0.1, 1.2],
+    "k": 10,
+    "num_candidates": 100
+  },
+  "_source": ["name", "date"]
+}
+----
+// TEST[setup:my_index]
+
+[[knn-search-api-request]]
+==== {api-request-title}
+
+`GET /<target>/_knn_search`
+
+`POST /<target>/_knn_search`
+
+[[knn-search-api-prereqs]]
+==== {api-prereq-title}
+
+* If the {es} {security-features} are enabled, you must have the `read`
+<<privileges-list-indices,index privilege>> for the target data stream, index,
+or alias.
+
+[[knn-search-api-desc]]
+==== {api-description-title}
+
+The kNN search API performs a k-nearest neighbor search on a
+<<dense-vector,`dense_vector`>> field. Given a query vector, it finds the _k_
+closest vectors and returns those documents as search hits. To be used in the
+kNN search API, the `dense_vector` field must
+<<index-vectors-knn-search, enable indexing>>.
+
+{es} uses the https://arxiv.org/abs/1603.09320[HNSW algorithm] to support
+efficient kNN search. Like most kNN algorithms, HNSW is an approximate method
+that sacrifices result accuracy for improved speed. This means the documents
+returned are not always the true _k_ closest neighbors.
+
+[[knn-search-api-path-params]]
+==== {api-path-parms-title}
+
+`<target>`::
+(Optional, string) Comma-separated list of data streams, indices, and aliases
+to search. Supports wildcards (`*`). To search all data streams and indices,
+use `*` or `_all`.
+
+NOTE: The kNN search API does not support remote indices (as in cross-cluster
+search).
+
+[role="child_attributes"]
+[[knn-search-api-query-params]]
+==== {api-query-parms-title}
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=routing]
+
+[role="child_attributes"]
+[[knn-search-api-request-body]]
+==== {api-request-body-title}
+
+`knn`::
+(Required, object) Defines the kNN query to run.
++
+.Properties of `knn` object
+[%collapsible%open]
+====
+`field`::
+(Required, string) The name of the vector field to search against.
+
+`query_vector`::
+(Required, array of floats) The query vector.
+
+`k`::
+(Required, integer) The final number of nearest neighbors to return as top
+hits. This value must be less than `num_candidates`.
+
+`num_candidates`::
+(Required, integer) The number of nearest neighbor candidates to consider per
+shard. Increasing `num_candidates` tends to improve the accuracy of the final
+`k` results. This value cannot exceed 10,000.
```

Comment:
We don't really describe the high-level process for gathering candidates from each shard and then collating the results. IIUC that and HSNW are the two main factors that can affect accuracy.

Should we save that for the high-level guide or mention in the description?

```suggestion
(Required, integer) Number of nearest neighbor candidates to consider per shard.
Cannot exceed `10000`. A higher value improves the accuracy of the results at
the cost of a slower search speed.
```

Reply:
I agree with @jrodewig it would be nice to add more details here how this `num_candidates` works, for example: "{es}  collects from each shard  the top `num_candidates` results, and then merges collected from the shards results to get the top k results.  Increasing `num_candidates` tends to improve the accuracy of the final `k` results..."

Reply:
This is a nice idea I added the sentence " {es} collects `num_candidates` results from each shard, then merges
them to find the top `k` results". In the high level guide, I think we can step through how the kNN search is executed in more detail.